students/OwenGaustad/Lab4/TurnSignal/TurnSignal.c

```diff
@@ -0,0 +1,184 @@
+
+first delay is roughly 100ms the second takes place after they are all turned on and is 200ms
+***********************************************************************************************/
+rightTurn() {
+  LED_PORT = 0b00010000;
+  for(led = 4; led >= 0; led--) {
+    toggleBit(LED_PORT, led);
+    variableDelay(100);
+  }
+  variableDelay(200);
+}
+/***********************************************
+
+This method checks to see if the left button
+
+was clicked then returns 1 if yes and 0 if no
+***********************************************/
+leftBtn() {
+  if(bit_is_clear(BUTTON_PIN, BUTTON2)){
+    return 1;
+  }
+  else {
+    return 0;  
+  }
+}
+/***********************************************
+
+This method checks to see if the right button 
+
+was clicked then returns 1 if yes and 0 if no 
+***********************************************/
+rightBtn() {
+  if(bit_is_clear(BUTTON_PIN, BUTTON)){
+    return 1;
+  }
+  else {
+    return 0;  
+  }
+}
+/*******************************************************************************
+
+This method takes an integer value and stores it as a temporary variable 
+
+time Then the delay lasts for the input delay time in milliseconds              
+*******************************************************************************/
+variableDelay(time) {
+  for (delay = 0; delay <= time; delay++) {
+    _delay_ms(1);
+    if (leftBtn() == 1 | rightBtn() == 1){
+      break;
+    }
+  }
+}
+/**************
+* Main Method *
+**************/
+int main(void){
+  uint8_t bita;                       //unused
+  uint8_t bitb;                       //unused
+  BUTTON_DDR &= ~(1 << BUTTON);       //activates input
+  BUTTON_PORT |= (1 << BUTTON);       //activates resistor
+  BUTTON_DDR &= ~(1 << BUTTON2);      //activates input
+  BUTTON_PORT |= (1 << BUTTON2);      //activates resistor
+  LED_DDR = 0xff;                     // All output
+
+  /****************************
+  * This Loops the LED states *
+  ****************************/
+  while (1){
+    /**********************************
+    * This Statement sets the LED     *
+    * state to all pins off           *
+    **********************************/
+    if (offState == 1) {
+      while (rightBtn() == 1 || leftBtn() == 1){
+        _delay_us(5);
+      }
+      while(rightBtn()==0 && leftBtn()==0 ){
+        LED_PORT = 0b00000000;
+      }
+      offState = 0;
+    }
+    /************************************
+    * This Statement sets the LED       *
+    * state to the right turn method    *
+    * if the right button is hit a      *
+    * second time the loop ends and     *
+    * all LED's are turned off but if   *
+    * the left button is hit it         *
+    * merely activates the other lights.* 
+    ************************************/
+    else if (rightBtn())  {
+      while(rightBtn()==1){
+        _delay_us(5);
+      }
+      while(rightBtn()==0 && leftBtn()==0 ){
```

Comment:
Man, there are a lot of button checks at different levels.
Something to improve would be to figure out a way to check the button in one place and have that one check execute whatever needs to be done. I'm imagining fringe situations where the delay function checked the button, then when the code gets here the button is in a different state and things get weird.